pkg/reconcilers/external/kubeconfig.go

```diff
@@ -0,0 +1,304 @@
+			Name:     "cluster-admin",
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				Kind:      "ServiceAccount",
+				Name:      saName,
+				Namespace: saNamespace,
+			},
+		},
+	}
+
+	// Create the ClusterRoleBinding
+	_, err = clientset.RbacV1().ClusterRoleBindings().Create(context.TODO(), clusterRoleBinding, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func reconcileAdoptedServiceAccount(saClient v1.ServiceAccountInterface, saName string) error {
+
+	// Check if the ServiceAccount already exists
+	_, err := saClient.Get(context.TODO(), saName, metav1.GetOptions{})
+	if err == nil {
+		return nil
+	}
+	if !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	// Define a new ServiceAccount
+	sa := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: saName,
+		},
+	}
+
+	// Create the ServiceAccount
+	_, err = saClient.Create(context.TODO(), sa, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func requestTokenWithExpiration(saClient v1.ServiceAccountInterface, saName string, expirationSeconds *int64) (string, error) {
+	if expirationSeconds == nil {
+		expirationSeconds = &defaultAdoptedTokenExpirationSeconds
+	}
+
+	tokenRequest := &authenticationv1.TokenRequest{
+		Spec: authenticationv1.TokenRequestSpec{
+			Audiences:         []string{},
+			ExpirationSeconds: expirationSeconds,
+		},
+	}
+
+	tokenResponse, err := saClient.CreateToken(context.TODO(), saName, tokenRequest, metav1.CreateOptions{})
+	if err != nil {
+		return "", err
+	}
+	return tokenResponse.Status.Token, nil
+}
+
+func createNewKubeConfig(bootstrapConfig *api.Config, token string, hcp *tenancyv1alpha1.ControlPlane) ([]byte, error) {
+	context := bootstrapConfig.Contexts[bootstrapConfig.CurrentContext]
+
+	cluster := bootstrapConfig.Clusters[context.Cluster]
+	if cluster == nil {
+		return nil, fmt.Errorf("invalid cluster name %s for adoped cluster", context.Cluster)
+	}
+
+	kubeConfig := api.NewConfig()
+
+	kubeConfig.Clusters[context.Cluster] = &api.Cluster{
+		Server:                   cluster.Server,
+		CertificateAuthorityData: cluster.CertificateAuthorityData,
+	}
+
+	kubeConfig.AuthInfos[hcp.Name] = &api.AuthInfo{
+		Token: token,
+	}
+
+	kubeConfig.Contexts[hcp.Name] = &api.Context{
+		Cluster:  context.Cluster,
+		AuthInfo: hcp.Name,
+	}
+	kubeConfig.CurrentContext = hcp.Name
+
+	newKubeConfig, err := clientcmd.Write(*kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return newKubeConfig, nil
+}
+
+func (r *ExternalReconciler) ReconcileKubeconfigSecret(ctx context.Context, cp tenancyv1alpha1.ControlPlane, kubeconfig []byte) error {
+	namespace := util.GenerateNamespaceFromControlPlaneName(cp.Name)
+
+	// create kubeconfig secret
+	kubeConfigSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      util.AdminConfSecret,
+			Namespace: namespace,
+		},
+		Type: corev1.SecretTypeOpaque,
+		Data: map[string][]byte{util.KubeconfigSecretKeyInCluster: kubeconfig},
```

Comment:
Is `util.KubeconfigSecretKeyInCluster` the right key (and the only right key)? Isn't this kubeconfig usable from anywhere?